src/combinations_with_replacement.rs

```diff
@@ -106,6 +106,24 @@ where
         Some(self.pool.get_at(&self.indices))
     }
 
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        if self.first {
+            // In empty edge cases, stop iterating immediately
+            if !(self.indices.is_empty() || self.pool.get_next()) {
+                return None;
+            }
+            self.first = false;
+        } else if self.increment_indices() {
+            return None;
+        }
+        for _ in 0..n {
+            if self.increment_indices() {
+                return None;
+            }
+        }
```

Comment:
Beneficial to fold the `else` branch into a `for _ in 0..=n`? Looks simpler imho. (I recognize that this would somewhat alter the symmetry between `nth` and `next`: `next` would initializate and `increment_indices` once, whereas `nth` would initialize and `increment_indices` n times.)

Reply:
`0..n` makes more sense when `self.first` is true, and `0..=n` otherwise.
I guess I could `let times = if first {... n} else { n+1 }; for _ in 0..times {...}`

Reply:
Sorry, I got lost with the `else`. I think yours is good.

Reply:
Currently benchmarking and I think I get slightly better results with `if first {n} else {n+1}`.

Reply:
I committed the change, edited the benchmarks in the main message. `.nth(0)` is not a bit slower anymore and other `.nth` are very slightly faster (it's necessarily small compared to removing unnecessary allocations).
I guess the `n+1` would be an issue for `.nth(usize::MAX)` but I can live with that.